### PR TITLE
Rewrite algorithm for finding start/end for 1x60 spawnpoints

### PR DESCRIPTION
**Pull request by trew**
Several of my instances stopped working because they got stuck in a CPU intensive loop. This turned out to be the algorithm for finding start/end times for 1x60 spawnpoints. I rewrote it from scratch and also added unit tests to the project.

## Description
Now it will loop through all pokemons that were sighted 2 or more times (for 1x60 spawnpoints, we can't determine start/end times with pokemons that were only scanned once). This way we can for each found pokemon expand the start/end window and that way find the range in O(n).

Unit tests were added and as a result the args variable had to be moved from the root level of imports.

## Motivation and Context
Previous algorithm was O(n^2) and that became a huge problem for n=1000+.

## How Has This Been Tested?
Local instances with -st 16.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pogom/models.py [hunk 1]
```diff
@@ -1269,66 +1283,63 @@ def classify(cls, sp, scan_loc, now_secs, sighting=None):
...
+                earliest_time = try_earliest
+            if try_latest > latest_time:
+                latest_time = try_latest
+
+        appear_time = earliest_time
+        despawn_time = latest_time
+
+        sp['latest_seen'] = despawn_time
+        sp['earliest_unseen'] = appear_time
+        log.info('1x60: appear %d, despawn %d, duration: %d min', appear_time, despawn_time, ((despawn_time - appear_time) % 3600) / 60)
```
Comment (sebastienvercammen) [inline pogom/models.py]: It's better to move this to `.debug()` for consistency.

File: pogom/models.py [hunk 2]
```diff
@@ -71,7 +71,7 @@ class BaseModel(flaskDb.Model):
     @classmethod
     def get_all(cls):
         results = [m for m in cls.select().dicts()]
-        if args.china:
+        if get_args().china:
```
Comment (sebastienvercammen) [inline pogom/models.py]: As mentioned on Discord, avoid doing this. For clarity, it should become at least `args = get_args()` and then use `args` later on.

For dependency injection, moving `args` to the first argument of the functions instead is even better, this is my preferred approach.

File: pogom/customLog.py [hunk 3]
```diff
@@ -12,7 +12,8 @@
 
 
 def printPokemon(id, lat, lng, itime):
-    if get_args().display_in_console:
+    args = get_args()
```
Comment (sebastienvercammen) [inline pogom/customLog.py]: For dependency injection, moving args to the first argument of the functions instead is even better, which is my preferred approach. This is even already being done in some places.

Perhaps it's best to discuss this now so we can stay consistent,

File: pogom/models.py (hunk 1 above)
Comment (trew) [inline pogom/models.py]: This was from earlier code, but I could of course change that.

File: pogom/customLog.py (hunk 3 above)
Comment (trew) [inline pogom/customLog.py]: I agree.

File: pogom/models.py (hunk 1 above)
Comment (sebastienvercammen) [inline pogom/models.py]: Yup that would be great, it would be more consistent with the other information being logged :)

File: pogom/models.py [hunk 4]
```diff
@@ -1316,7 +1326,7 @@ def classify(cls, sp, scan_loc, now_secs, sighting=None):
         # An hour minus the largest gap in minutes gives us the duration the
         # spawn was there.  Round up to the nearest 15 minute interval for our
         # current best guess duration.
-        duration = (int((59 - max_gap / 60.0) / 15) + 1) * 15
+        duration = (int((60.0 - (max_gap / 60.0)) / 15) + 1) * 15
```
Comment (j16sdiz) [inline pogom/models.py]: when max_gap == 0, duration is 75. Is this possible ?

Comment (trew) [inline pogom/models.py]: I really hope this method never gets called if max_gap == 0. That'd mean the scheduler was done scanning the spawnpoint and wants to classify the type, but the only times it has scanned at is for example 10:00, 11:00, 12:00 etc, aka max_gap == 0.

**Comment by j16sdiz** [comment]
1. How to run the test cases?

2. I have compared with the old code. Seems the old code was generating wrong disappear time.

**Comment by j16sdiz** [comment]
+1 for correct spawn time for 1x60.  Existing code just give wrong time.

**Comment by trew** [comment]
Tests can be run from the root directory with
`python -m unittest discover tests`

File: pogom/models.py [hunk 5]
```diff
@@ -1293,9 +1297,15 @@ def classify(cls, sp, scan_loc, now_secs, sighting=None):
...
                         .dicts())
 
+        cls.classify_sp(query, sp, sighting)
+
+    @classmethod
+    def classify_sp(cls, query, sp, sighting):
         if sighting:
             query.append(sighting)
 
+        query.sort(key=lambda x: x['scan_time'], reverse=True)
```
Comment (j16sdiz) [inline pogom/models.py]: I am wonder if this should be sorted by `date_secs(x['scan_time'])` instead. 

I got this misclassified as 'hsss':
```
  add_query_entry(query, '1', 18, 0, 0)                                                                                                                                                                                          
  add_query_entry(query, '1', 18, 12, 59)                                                                                                                                                                                        
  add_query_entry(query, '1', 18, 18, 59)                                                                                                                                                                                        
  add_query_entry(query, '1', 18, 55, 59)                                                                                                                                                                                        
  
  add_query_entry(query, '2', 20, 30, 0)                                                                                                                                                                                         
  add_query_entry(query, '2', 20, 46, 59)          
```

Comment (trew) [inline pogom/models.py]: Changing to date_secs there doesn't matter. date_secs is used later for determining the max_gap, which is used to determine the spawnpoint type, and this is where the bug lies.

**Comment by j16sdiz** [comment]
this pull is conflict with the current head.

**Comment by sebastienvercammen** [comment]
@j16sdiz The author knows, it says so right at the bottom of the page. Please use PR comments only for constructive feedback.

**Comment by trew** [comment]
~Note: This should not be merged yet as it still contains bugs.~

**Comment by sLoPPydrive** [comment]
Does this PR also re-scan already found 60min spawns? In the past days I encountered some spawns my map had marked as 60 min spawns and which had Pokemon I'd like to catch. When I arrived at the spawn the Pokemon wasn't there. Instead another Pokemon took its place, so I strongly believe the spawn was classified incorrectly and needs to be re-scanned.

**Comment by trew** [comment]
This PR does not reclassify existing spawnpoints. I've yet to look into how that can be achieved though.

**Comment by j16sdiz** [comment]
Under this pull, I got a 30minutes spawn misclassified as 60:


This is the detection data sort by time of hour.
```
mysql> select * from spawnpointdetectiondata where spawnpoint_id = '3404089273b' order by right(scan_time ,5);
+-----------------------------------+------------------------------+---------------+---------------------+----------+
| id                                | encounter_id                 | spawnpoint_id | scan_time           | tth_secs |
+-----------------------------------+------------------------------+---------------+---------------------+----------+
| MTYxODM5MjEwNTY3MTU3MTA1Nzk=_49   | MTYxODM5MjEwNTY3MTU3MTA1Nzk= | 3404089273b   | 2017-02-10 05:00:49 |     NULL |
| MTYxODM5MjEwNTY3MTU3MTA1Nzk=_158  | MTYxODM5MjEwNTY3MTU3MTA1Nzk= | 3404089273b   | 2017-02-10 05:02:38 |     NULL |
| MTYxOTg4MTY1ODY2NzM3NTQxMzY=_205  | MTYxOTg4MTY1ODY2NzM3NTQxMzY= | 3404089273b   | 2017-02-11 08:03:26 |     NULL |
| OTA1MDcwNzU1OTU0NzQzMDU5MQ==_209  | OTA1MDcwNzU1OTU0NzQzMDU5MQ== | 3404089273b   | 2017-02-10 17:03:30 |     NULL |
| OTA1MDcwNzU1OTU0NzQzMDU5MQ==_211  | OTA1MDcwNzU1OTU0NzQzMDU5MQ== | 3404089273b   | 2017-02-10 17:03:31 |     NULL |
| OTA1MDcwNzU1OTU0NzQzMDU5MQ==_251  | OTA1MDcwNzU1OTU0NzQzMDU5MQ== | 3404089273b   | 2017-02-10 17:04:11 |     NULL |
| OTA1MDcwNzU1OTU0NzQzMDU5MQ==_289  | OTA1MDcwNzU1OTU0NzQzMDU5MQ== | 3404089273b   | 2017-02-10 17:04:49 |     NULL |
| MTYxODM5MjEwNTY3MTU3MTA1Nzk=_325  | MTYxODM5MjEwNTY3MTU3MTA1Nzk= | 3404089273b   | 2017-02-10 05:05:25 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_335  | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 06:05:36 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_341  | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 06:05:41 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_369  | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 06:06:10 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_375  | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 06:06:15 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_376  | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 06:06:17 |     NULL |
| MTYxOTg4MTY1ODY2NzM3NTQxMzY=_414  | MTYxOTg4MTY1ODY2NzM3NTQxMzY= | 3404089273b   | 2017-02-11 08:06:54 |     NULL |
| MTYxOTg4MTY1ODY2NzM3NTQxMzY=_472  | MTYxOTg4MTY1ODY2NzM3NTQxMzY= | 3404089273b   | 2017-02-11 08:07:52 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_488  | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 06:08:09 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_510  | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 06:08:30 |      587 |
| MTMyMzQzODgyNDMzNzU3OTIwMDQ=_520  | MTMyMzQzODgyNDMzNzU3OTIwMDQ= | 3404089273b   | 2017-02-10 18:08:41 |      587 |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_537  | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 06:08:58 |     NULL |
| NjQ3NDg4MzQ0MzAwMDM0MTIzMQ==_2397 | NjQ3NDg4MzQ0MzAwMDM0MTIzMQ== | 3404089273b   | 2017-02-11 08:39:58 |     NULL |
| NjQ3NDg4MzQ0MzAwMDM0MTIzMQ==_2408 | NjQ3NDg4MzQ0MzAwMDM0MTIzMQ== | 3404089273b   | 2017-02-11 08:40:09 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_2425 | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 05:40:26 |     NULL |
| MTMyMzQzODgyNDMzNzU3OTIwMDQ=_2448 | MTMyMzQzODgyNDMzNzU3OTIwMDQ= | 3404089273b   | 2017-02-10 17:40:48 |     NULL |
| NjQ3NDg4MzQ0MzAwMDM0MTIzMQ==_2455 | NjQ3NDg4MzQ0MzAwMDM0MTIzMQ== | 3404089273b   | 2017-02-11 08:40:56 |     NULL |
| NjQ3NDg4MzQ0MzAwMDM0MTIzMQ==_2582 | NjQ3NDg4MzQ0MzAwMDM0MTIzMQ== | 3404089273b   | 2017-02-11 08:43:03 |     NULL |
| MTMyMzQzODgyNDMzNzU3OTIwMDQ=_2671 | MTMyMzQzODgyNDMzNzU3OTIwMDQ= | 3404089273b   | 2017-02-10 17:44:31 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_2706 | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 05:45:06 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_2730 | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 05:45:30 |     NULL |
| NjQ3NDg4MzQ0MzAwMDM0MTIzMQ==_2761 | NjQ3NDg4MzQ0MzAwMDM0MTIzMQ== | 3404089273b   | 2017-02-11 08:46:02 |     NULL |
| NjQ3NDg4MzQ0MzAwMDM0MTIzMQ==_2763 | NjQ3NDg4MzQ0MzAwMDM0MTIzMQ== | 3404089273b   | 2017-02-11 08:46:03 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_2799 | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 05:46:39 |     NULL |
| MTMyMzQzODgyNDMzNzU3OTIwMDQ=_2810 | MTMyMzQzODgyNDMzNzU3OTIwMDQ= | 3404089273b   | 2017-02-10 17:46:50 |     NULL |
| MTMyMzQzODgyNDMzNzU3OTIwMDQ=_2840 | MTMyMzQzODgyNDMzNzU3OTIwMDQ= | 3404089273b   | 2017-02-10 17:47:21 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_2894 | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 05:48:14 |     NULL |
| NjQ3NDg4MzQ0MzAwMDM0MTIzMQ==_2967 | NjQ3NDg4MzQ0MzAwMDM0MTIzMQ== | 3404089273b   | 2017-02-11 08:49:28 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_3017 | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 05:50:17 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_3016 | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 05:50:17 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_3060 | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 05:51:00 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_3059 | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 05:51:00 |     NULL |
| MTMyMzQzODgyNDMzNzU3OTIwMDQ=_3142 | MTMyMzQzODgyNDMzNzU3OTIwMDQ= | 3404089273b   | 2017-02-10 17:52:22 |     NULL |
| MTMyMzQzODgyNDMzNzU3OTIwMDQ=_3161 | MTMyMzQzODgyNDMzNzU3OTIwMDQ= | 3404089273b   | 2017-02-10 17:52:41 |     NULL |
| MTMyMzQzODgyNDMzNzU3OTIwMDQ=_3160 | MTMyMzQzODgyNDMzNzU3OTIwMDQ= | 3404089273b   | 2017-02-10 17:52:41 |     NULL |
| NjQ3NDg4MzQ0MzAwMDM0MTIzMQ==_3175 | NjQ3NDg4MzQ0MzAwMDM0MTIzMQ== | 3404089273b   | 2017-02-11 08:52:55 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_3184 | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 05:53:05 |     NULL |
| MTMyMzQzODgyNDMzNzU3OTIwMDQ=_3189 | MTMyMzQzODgyNDMzNzU3OTIwMDQ= | 3404089273b   | 2017-02-10 17:53:09 |     NULL |
| MTYxODM5MjEwNTY3MTU3MTA1Nzk=_3266 | MTYxODM5MjEwNTY3MTU3MTA1Nzk= | 3404089273b   | 2017-02-10 04:54:26 |     NULL |
| OTA1MDcwNzU1OTU0NzQzMDU5MQ==_3286 | OTA1MDcwNzU1OTU0NzQzMDU5MQ== | 3404089273b   | 2017-02-10 16:54:46 |     NULL |
| MTMyMzQzODgyNDMzNzU3OTIwMDQ=_3355 | MTMyMzQzODgyNDMzNzU3OTIwMDQ= | 3404089273b   | 2017-02-10 17:55:55 |     NULL |
| MTMyMzQzODgyNDMzNzU3OTIwMDQ=_3356 | MTMyMzQzODgyNDMzNzU3OTIwMDQ= | 3404089273b   | 2017-02-10 17:55:56 |     NULL |
| OTA1MDcwNzU1OTU0NzQzMDU5MQ==_3439 | OTA1MDcwNzU1OTU0NzQzMDU5MQ== | 3404089273b   | 2017-02-10 16:57:19 |     NULL |
| MTEwMDQwNzI1Mjk3MTgwOTQyNDg=_3452 | MTEwMDQwNzI1Mjk3MTgwOTQyNDg= | 3404089273b   | 2017-02-10 05:57:32 |     NULL |
| MTYxODM5MjEwNTY3MTU3MTA1Nzk=_3536 | MTYxODM5MjEwNTY3MTU3MTA1Nzk= | 3404089273b   | 2017-02-10 04:58:56 |     NULL |
| MTMyMzQzODgyNDMzNzU3OTIwMDQ=_3590 | MTMyMzQzODgyNDMzNzU3OTIwMDQ= | 3404089273b   | 2017-02-10 17:59:51 |     NULL |
+-----------------------------------+------------------------------+---------------+---------------------+----------+
```

```
mysql> select * from spawnpoint where id='3404089273b';
+-------------+------------------+------------------+---------------------+------+-------+--------------+-------------+-----------------+
| id          | latitude         | longitude        | last_scanned        | kind | links | missed_count | latest_seen | earliest_unseen |
+-------------+------------------+------------------+---------------------+------+-------+--------------+-------------+-----------------+
| 3404089273b | 22.4335169991477 | 114.203895144243 | 2017-02-11 08:53:07 | ssss | +++-  |           84 |         587 |             587 |
+-------------+------------------+------------------+---------------------+------+-------+--------------+-------------+-----------------+
1 row in set (0.00 sec)
```

It should be 30minutes from the detection data above : no spawn from xx:08:58 to xx:39:58, 
Although the entry at "2017-02-10 06:08:58" is really strange.

**Comment by trew** [comment]
That's too bad. Oh well. I don't have time or interest to continue this PR since I neither maintain a map nor play the game any more.

**Comment by smint86** [comment]
I was eagerly waiting for this and planned a review this weekend. Maybe anyone else will pick this up. It looked nice so far.

**Pull request closed without merging**